packages/circuits/test/utils.ts

```diff
@@ -452,123 +439,378 @@ const genProcessAttestationsCircuitInput = (
+            if (hashChains[epochKey] === undefined)
+                hashChains[epochKey] = BigInt(0)
+            hashChains[epochKey] = crypto.hashLeftRight(
+                a.hash(),
+                hashChains[epochKey]
+            )
+        }
+    }
+
+    for (const key in hashChains) {
+        const sealedHashChain = crypto.hashLeftRight(1, hashChains[key])
+        fromEpochTree.update(BigInt(key), sealedHashChain)
+    }
+
+    const epochTreeRoot = fromEpochTree.root
+    const epochKeyPathElements: any[] = []
 
-    const hashChainResults: BigInt[] = []
-    // Begin generating and processing attestations
+    // start transition proof
+    const {
+        circuitInputs: startTransitionCircuitInputs,
+        blindedUserState: startTransitionBlindedUserState,
+    } = genStartTransitionCircuitInput(
+        id,
+        epoch,
+        fromNonce,
+        fromEpochUserStateTree.root,
+        GSTreeProof
+    )
+
+    // process attestation proof
+    const processAttestationCircuitInputs: any[] = []
+    const fromNonces: number[] = [fromNonce]
+    const toNonces: number[] = []
+    const hashChainStarter: BigInt[] = []
+    const blindedUserState: BigInt[] = [startTransitionBlindedUserState]
+    const blindedHashChain: BigInt[] = []
+    let reputationRecords = {}
+    const selectors: number[] = []
+    const attesterIds: string[] = []
+    const oldPosReps: string[] = [],
+        oldNegReps: string[] = [],
+        oldGraffities: string[] = [],
+        oldSignUps: string[] = []
+    const posReps: string[] = [],
+        negReps: string[] = [],
+        graffities: string[] = [],
+        overwriteGraffities: any[] = [],
+        signUps: string[] = []
+    const finalBlindedUserState: BigInt[] = []
+    const finalUserState: BigInt[] = [intermediateUserStateTreeRoots[0]]
+    const finalHashChain: BigInt[] = []
     for (let nonce = 0; nonce < NUM_EPOCH_KEY_NONCE_PER_EPOCH; nonce++) {
-        // Each epoch key has `ATTESTATIONS_PER_EPOCH_KEY` of attestations so
-        // interval between starting index of each epoch key is `ATTESTATIONS_PER_EPOCH_KEY`.
         const epochKey = genEpochKey(
             id.identityNullifier,
-            epoch,
+            fromEpoch,
             nonce,
             EPOCH_TREE_DEPTH
         )
-        const hashChainResult = crypto.genRandomSalt()
+        let currentHashChain: BigInt = BigInt(0)
+
+        // Blinded user state and hash chain of the epoch key
+        toNonces.push(nonce)
+        hashChainStarter.push(currentHashChain)
+
+        // Attestations
+        const attestations: Attestation[] = attestationsMap[epochKey.toString()]
+
+        // TODO: update attestation types
+        for (let i = 0; i < attestations.length; i++) {
+            // Include a blinded user state and blinded hash chain per proof
+            if (
+                i &&
+                i % NUM_ATTESTATIONS_PER_PROOF == 0 &&
+                i != NUM_ATTESTATIONS_PER_PROOF - 1
```

Comment:
If i % NUM_ATTESTATIONS_PER_PROOF and i > 0, this mean `i` always different NUM_ATTESTATIONS_PER_PROOF - 1 I think.